components/raftstore/src/store/worker/pd.rs

```diff
@@ -1080,13 +1080,72 @@ where
                     }
                     if resp.get_require_detailed_report() {
                         info!("required to send detailed report in the next heartbeat");
-                        let task = Task::StoreHeartbeat {
-                            stats: stats_copy,
-                            store_info,
-                            send_detailed_report: true,
-                        };
-                        if let Err(e) = scheduler.schedule(task) {
-                            error!("notify pd failed"; "err" => ?e);
+                        let mut commit_indices = HashMap::default();
+                        store_info
+                            .kv_engine
+                            .scan_cf(
+                                CF_RAFT,
+                                keys::REGION_META_MIN_KEY,
+                                keys::REGION_META_MAX_KEY,
+                                false,
+                                |key, value| {
+                                    let (_, suffix) = box_try!(keys::decode_region_meta_key(key));
+                                    if suffix != keys::REGION_STATE_SUFFIX {
+                                        return Ok(true);
+                                    }
+
+                                    let mut region_local_state = RegionLocalState::default();
+                                    region_local_state.merge_from_bytes(value)?;
+                                    if region_local_state.get_state() == PeerState::Tombstone {
+                                        return Ok(true);
+                                    }
+                                    let apply_state: RaftApplyState = match store_info
+                                        .kv_engine
+                                        .get_msg_cf(
+                                            CF_RAFT,
+                                            &keys::apply_state_key(
+                                                region_local_state.get_region().get_id(),
+                                            ),
+                                        )
+                                        .unwrap()
+                                    {
+                                        None => return Ok(true),
+                                        Some(value) => value,
+                                    };
+                                    if apply_state.get_applied_index()
+                                        != apply_state.get_last_commit_index()
+                                    {
+                                        commit_indices.insert(
```

Comment:
no need to check here. broadcast the message to every peer fsm. And in each peer fsm, you can get the apply index and commit index from memory.